base/testfiles-ltcmd/ltcmd001.tlg

```diff
@@ -700,14 +668,14 @@ TEST 17: (ab)using xparse commands in csnames
 l. ...  }
 The control sequence marked <to be read again> should
 not appear between \csname and \endcsname.
-\test- \xparse function is not expandable \bool_set_false:N \l__cmd_environment_bool \__cmd_start_aux:NNnnnn \foo code {D...}{\__cmd_grab_D:w ..}{{.}}{}....\cs_end: 
+\test- \xparse function is not expandable \bool_set_false:N \l__cmd_environment_bool \__cmd_start_aux:NNnnnn \foo  \foo code {D...}{\__cmd_grab_D:w ..}{{.}}{}....\cs_end: 
```

Comment:
This change (`\foo␣code {D...}` -> `\foo␣ \foo␣code {D...}`) took place because previously for expandable command `\cmd`, `\cmd␣` is also defined (and if that command has both short and long arguments, `\cmd␣␣` is defined in addition). All happen in `\__cmd_declare_cmd_code_expandable:Nnn`.

Prior to this PR, in `ltcmd001.lvt` lines 363 and 368--369,
- first `\DeclareDocumentCommand { \foo } { m m } { #1-#2 }`  defines `\foo␣` to `macro:#1#2->#1{#2}`,
- then `\DeclareDocumentCommand { \foo } { D... } { test-#1 }` leaves `\foo␣` intact,
- finally in `\TYPE { \exp_not:c { test- \foo .... } }`,  `\foo␣ \foo␣code {D...}` expands to `\foo␣code {D...}`, which is then written to log.
 
Now `\foo␣` is made `\relax` by `\DeclareDocumentCommand { \foo } { m m } { #1-#2 }` (see `\__cmd_declare_cmd_internal:Nnnn`), hence left in log.

https://github.com/latex3/latex2e/blob/0f40cbb8ea0c43592db9d1c96dbb7575166b4a7b/base/testfiles-ltcmd/ltcmd001.lvt#L361-L375

Several questions are involved:
- Can optimized expandable commands leave `\foo␣` `\relax`ed?
- Should `\(Renew|Declare)DocumentCommand` and friends clear command internals before redefinition? See also #1191.
- In what extent should (ab)using document commands in csnames (name of the test these log lines correspond to) be guarded and tested?
  Do optimized commands need to add a such guard? This may lead to a new `\__cmd_start_optimized_expandable:` and the old `\__cmd_start_optimized:` containing (full expansion of) something like `\exp_not:c { xparse~function~is~not~expandable }`.

These are all not firmly related to current PR hence I think they can be discussed and probably resolved in separate issue(s) and PR(s).

Reply:
I've moved part of the setup as it avoids creating the csname before it's required. That's of course separate from the other queries you raise.

Reply:
> * In what extent should (ab)using document commands in csnames (name of the test these log lines correspond to) be guarded and tested?
>   Do optimized commands need to add a such guard? This may lead to a new `\__cmd_start_optimized_expandable:` and the old `\__cmd_start_optimized:` containing (full expansion of) something like `\exp_not:c { xparse~function~is~not~expandable }`.

This one is not needed, as document commands having all-`m` arguments always start with `\__cmd_start_expandable:nNNNNn`, never `\__cmd_start:nNNnnn`, see for example #1013.